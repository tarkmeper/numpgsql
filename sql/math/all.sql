select aall('{}'::integer[]);
select aall('{1}'::integer[]);
select aall('{0}'::integer[]);
select aall('{1,3,5,1}'::integer[]);
select aall('{1,3,0,1}'::integer[]);
select aall('{{1,3},{0,1}}'::integer[]);

select aall('{true}'::boolean[]);
select aall('{false}'::boolean[]);

