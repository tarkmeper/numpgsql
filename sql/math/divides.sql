SELECT divide('{3,4}'::integer[], '{1,2}'::integer[]);

--Div by 0
SELECT divide('{3,4}'::integer[], '{0,0}'::integer[]);

--Single value
SELECT '{{6,9},{3,4}}'::integer[] / 3;
SELECT divide('{3,4}'::integer[], 0);
