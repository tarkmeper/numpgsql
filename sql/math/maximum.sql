SELECT maximum('{1,4,5,6,7}'::integer[],5);
