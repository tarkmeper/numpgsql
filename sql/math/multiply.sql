SELECT multiply('{1,2}'::integer[], '{3,4}'::integer[]);

--Single value
SELECT '{{1,2},{3,4}}'::integer[] * 3;
