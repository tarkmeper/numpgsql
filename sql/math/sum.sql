SELECT '{1,2}'::integer[] + '{3,4}'::integer[]
