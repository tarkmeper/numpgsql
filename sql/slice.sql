-- Integer arrays
SELECT ARRAY[1,2,3,4,5,6] @ ARRAY[1];
SELECT ARRAY[1,2,3,4,5,6] @ ARRAY[-1];
SELECT ARRAY[1,2,3,4,5,6] @ ARRAY[1,1,1,-1,-1];

--Other Types
SELECT ARRAY['asd', 'fds', 'wef'] @ ARRAY[2, -2, 2];

-- out of bounds
SELECT ARRAY[1,2,3,4,5,6] @ ARRAY[10];
SELECT ARRAY[1,2,3,4,5,6] @ ARRAY[0];
SELECT ARRAY[1,2,3,4,5,6] @ ARRAY[-7];

