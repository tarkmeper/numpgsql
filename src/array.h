extern "C"  {
        #include "postgres.h"
        #include "catalog/pg_type.h"
        #include "utils/array.h"
        #include "fmgr.h"
}
#include <vector>
#include <tuple>


inline static size_t arr_cnt(const ArrayType* arr) {
	size_t result = 1;
	for (int i = 0; i < ARR_NDIM(arr); ++i){
		result *= ARR_DIMS(arr)[i];
	}
	return result;
}

//Return values - TODO rename function this was poorly named.
template<class T> inline Datum _ret(const T& val) { T::unimplemented_function; return Datum(); }
template<> inline Datum _ret<double>(const double& val) { PG_RETURN_FLOAT8(val); }
template<> inline Datum _ret<float>(const float& val) { PG_RETURN_FLOAT4(val); }
template<> inline Datum _ret<short>(const short& val) { PG_RETURN_INT16(val); }
template<> inline Datum _ret<int>(const int& val) { PG_RETURN_INT32(val); }
template<> inline Datum _ret<long>(const long& val) { PG_RETURN_INT64(val); }
template<> inline Datum _ret<bool>(const bool& val) { PG_RETURN_BOOL(val); }

//OID types
template<class T> inline Oid oid_type() { return T::unimplemented_feature; }
template<> inline Oid oid_type<float>() { return FLOAT4OID; }
template<> inline Oid oid_type<double>() { return FLOAT8OID; }
template<> inline Oid oid_type<short>() { return INT2OID; }
template<> inline Oid oid_type<int>() { return INT4OID; }
template<> inline Oid oid_type<long>() { return INT8OID; }
template<> inline Oid oid_type<bool>() { return BOOLOID; }

//Alignment codes
template<class T> char oid_align() { return T::unimplemented_feature; }
template<> inline char oid_align<float>() { return 'i'; }
template<> inline char oid_align<double>() { return 'd'; }
template<> inline char oid_align<short>() { return 's'; }
template<> inline char oid_align<int>() { return 'i'; }
template<> inline char oid_align<long>() { return 'd'; }
template<> inline char oid_align<bool>() { return 'c'; }

//Get param values
template<class T> inline T get_param(PG_FUNCTION_ARGS, int p)  { T::unimplemented_function; }
template<> inline bool get_param<bool>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_BOOL(p); }
template<> inline float get_param<float>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_FLOAT4(p); }
template<> inline double get_param<double>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_FLOAT8(p); }
template<> inline short get_param<short>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_INT16(p); }
template<> inline int get_param<int>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_INT32(p); }
template<> inline long get_param<long>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_INT64(p); }
