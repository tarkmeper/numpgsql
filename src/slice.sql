-- This program is open source, licensed under the MIT License.
-- For license terms, see the LICENSE file.
--
-- Copyright (C) 2019: Mark Teper

CREATE FUNCTION slice(anyarray, integer[]) RETURNS anyarray LANGUAGE C STRICT IMMUTABLE LEAKPROOF COST 100 PARALLEL SAFE AS '$libdir/numpgsql', 'slice_idx';
CREATE FUNCTION slice(anyarray, boolean[]) RETURNS anyarray LANGUAGE C STRICT IMMUTABLE LEAKPROOF COST 100 PARALLEL SAFE AS '$libdir/numpgsql', 'slice_bit';

CREATE OPERATOR @ (LEFTARG = anyarray, RIGHTARG = integer[], PROCEDURE = slice);
CREATE OPERATOR @ (LEFTARG = anyarray, RIGHTARG = boolean[], PROCEDURE = slice);
