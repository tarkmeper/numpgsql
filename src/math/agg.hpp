#ifndef AGG_HPP
#define AGG_HPP
//TODO - compilation is slow and produces large binaries due to the debug symbols from BOOST.  Determine
//if tehre is some way to disable those symbols.
extern "C"  {
	#include "postgres.h"
	#include "catalog/pg_type.h"
	#include "utils/array.h"
	#include "fmgr.h"

}
#include "../array.h"

#include <algorithm>
#include <functional>
#include <boost/accumulators/numeric/functional/vector.hpp>

using namespace boost::numeric::operators;


#include <boost/accumulators/accumulators.hpp>

using namespace boost::accumulators;

//////////////////////////////////////////////////////////////////////////////////////
//  Array Aggregation Functions
//  Compute the aggregation across a single array perhaps in the future with support
//  for aggregating along an axis.
/////////////////////////////////////////////////////////////////////////////////////

//Create an aceclerator for the target function and apply on the dataset.
template<class V, class T> static Datum _apply(T* ptr, size_t n) {
  accumulator_set<T, features<V> > acc;
  std::for_each(ptr, ptr + n, std::ref(acc));
  return _ret(extract_result<V>(acc));
}

template<class V> static Datum _op(PG_FUNCTION_ARGS) {
    ArrayType* arr = PG_GETARG_ARRAYTYPE_P(0);
    Oid type  = ARR_ELEMTYPE(arr);
    const size_t n = arr_cnt(arr);
    void* ptr = ARR_DATA_PTR(arr);
    switch(type) {
      case BOOLOID: return _apply<V> ( (bool*)ptr, n);
      case FLOAT4OID: return _apply<V>( (float4*)ptr, n );
      case FLOAT8OID: return _apply<V>( (float8*)ptr, n );
      case INT2OID: return _apply<V>( (int16*)ptr, n );
      case INT4OID: return _apply<V>( (int32*)ptr, n );
      case INT8OID: return _apply<V>( (int64*)ptr, n );
      default:
        elog(ERROR, "Unsupported array type");
        PG_RETURN_NULL();
    }
}


//Define a function for each aggregation which internally has a switch between the basic types.
//This is needed.  Todo - investigate if it is possible to generate more of this using C++ templates
//where the debuging is slightly easier then C macros.
#define ARRAY_AGGREGATE_FNC(NAME, OP)  \
  extern "C" { PG_FUNCTION_INFO_V1(NAME); Datum NAME(PG_FUNCTION_ARGS); } \
  Datum NAME(PG_FUNCTION_ARGS) { return _op<OP>(fcinfo); }


//////////////////////////////////////////////////////////////////////////////////////
//  SQL Aggregation Functions
//  Compute the aggregation across multiple arrays using an SQL group by clause.
//  Normally we would expect that all aggregartes would have both.
/////////////////////////////////////////////////////////////////////////////////////

struct AggInternal  {
  size_t cnt;
  void*  acc;
  ArrayType* (*fnc)(AggInternal*);
};

template<class V, class T> ArrayType* to_array(AggInternal* a) {
    typedef accumulator_set< std::vector<T>, features<V> > Acc_T;

    if (a->cnt == 0) {
      return construct_empty_array(FLOAT4OID);
    } else {
      Acc_T* acc = (Acc_T*)(a->acc);

      //Figure out result type automatically
      auto v = extract_result<V>(*acc);
      Datum* arr = (Datum*)palloc(a->cnt * sizeof(Datum));
      for (unsigned int i = 0; i < a->cnt; ++i) {
        arr[i] = Float4GetDatum((float)v[i]);
      }

      //Now delete the aggregator and set the value to nullptr.
      acc->~Acc_T();
      a->acc = nullptr;

      return construct_array(arr, a->cnt, FLOAT4OID, 4, true, 'i');
    }
}


//Apply takes the vector we have and assuming that the vector isn't empty and matches previous vector sizes
//will apply the accumulator for that vector.
template<class V, class T> Datum _sql_apply(MemoryContext& ctx, AggInternal* state, T* arr_ptr, const size_t n) {
  typedef accumulator_set< std::vector<T>, features<V> > Acc_T;

  if (!state) {
    //Use C memory allocator so that Postgres can manage the memory
    //critical that internal representation have trivial destructor
    //we need to force call the destructor on Acc_T.
    void *acc = MemoryContextAlloc(ctx, sizeof(Acc_T));
    acc = new (acc) Acc_T(std::vector<T>(n));

    state = (AggInternal*)MemoryContextAlloc(ctx, sizeof(AggInternal));
    state->cnt = n;
    state->acc = acc;
    state->fnc = &to_array<V,T>;
  }
  Acc_T* acc = (Acc_T*)(state->acc);

  //special case - if the array is empty allow the aggregation to continue
  //excluding it.
  if (n == 0) {
    PG_RETURN_POINTER(state);
  } else if ( n != state->cnt ) {
    elog(ERROR, "All vectors to be aggregated must have same number of elements");
    PG_RETURN_POINTER(state);
  }

  std::vector<T> v(arr_ptr, arr_ptr+n);
  (*acc)(v);
  PG_RETURN_POINTER(state);
};

template<class V, class T, bool ALL> struct Call {
  Datum operator()(MemoryContext&, AggInternal*, T* ptr, size_t n) { V::unimplimeneted_function; return Datum(); }
};

template<class V, class T> struct Call<V, T, true>{
  Datum operator()(MemoryContext& ctx, AggInternal* state, T* ptr, size_t n) {
    return _sql_apply<V>(ctx, state, ptr, n);
  }
};


template<class V, class T> struct Call<V, T, false>{
  Datum operator()(MemoryContext& ctx, AggInternal* state, T* ptr, size_t n) {
    std::vector<double> v(ptr, ptr+n);
    return _sql_apply<V>(ctx, state, v.data(), n);
  }
};



template<class V, bool ALL> static Datum _sql_op(PG_FUNCTION_ARGS) {
  MemoryContext ctx;
  AggInternal * state = nullptr;

  //Since we don't have strict on the array pointer could be
  //null
  if (PG_ARGISNULL(1)) {
     if (PG_ARGISNULL(0)) { PG_RETURN_NULL(); }
     else { PG_RETURN_POINTER(PG_GETARG_POINTER(0)); }
  }

  ArrayType* arr = PG_GETARG_ARRAYTYPE_P(1);
  Oid type  = ARR_ELEMTYPE(arr);
  const size_t n = arr_cnt(arr);
  void* ptr = ARR_DATA_PTR(arr);

  //Block calls when not in aggregation mode.
  if (!AggCheckCallContext(fcinfo, &ctx)) {
    elog(ERROR,  "Called in non-aggregate context");
    PG_RETURN_NULL();
  }

  //If we have a state variable already return that.
  if (!PG_ARGISNULL(0)) {
    state = (AggInternal*)PG_GETARG_POINTER(0);
  } else if ( n == 0 ) {
     // special case if we don't yet have a state, and the
     // size is 0 then return null
     PG_RETURN_NULL();
  }

  if (state && state->acc == nullptr) {
     elog(ERROR, "Aggregation has already been finalized - cannot continue aggregating");
     PG_RETURN_NULL();
  }

  switch(type) {
      case FLOAT4OID: return Call<V, float, ALL>()( ctx, state, (float4*)ptr, n);
      case FLOAT8OID: return Call<V, double, ALL>()( ctx, state, (float8*)ptr, n);
      case INT2OID: return Call<V, short, ALL>()( ctx, state, (int16*)ptr, n);
      case INT4OID: return Call<V, int, ALL>()( ctx, state, (int32*)ptr, n);
      case INT8OID: return Call<V, long, ALL>()( ctx, state, (int64*)ptr, n);
      default:
         elog(ERROR, "Unsupported Array Type");
         PG_RETURN_NULL();
  }
}


#define SQL_AGGREGATE_FNC(NAME, OP)  \
  extern "C" { PG_FUNCTION_INFO_V1(NAME); Datum NAME(PG_FUNCTION_ARGS); } \
  Datum NAME(PG_FUNCTION_ARGS) { return _sql_op<OP, true>(fcinfo); }

#define SQL_AGGREGATE_DBL_ONLY_FNC(NAME, OP)  \
  extern "C" { PG_FUNCTION_INFO_V1(NAME); Datum NAME(PG_FUNCTION_ARGS); } \
  Datum NAME(PG_FUNCTION_ARGS) { return _sql_op<OP, false>(fcinfo); }



#endif
