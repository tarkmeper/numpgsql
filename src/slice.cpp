extern "C"  {
	#include "postgres.h"
	#include "catalog/pg_type.h"
	#include "utils/array.h"
        #include <utils/lsyscache.h>
	#include "fmgr.h"
	#include "math.h"
}
#include "array.h"
#include <functional>
#include <tuple>

extern "C" { PG_FUNCTION_INFO_V1(slice_idx); Datum slice_idx(PG_FUNCTION_ARGS); }

Datum slice_idx(PG_FUNCTION_ARGS) {
  ArrayType* arr = PG_GETARG_ARRAYTYPE_P(0);
  ArrayType* idx_arr = PG_GETARG_ARRAYTYPE_P(1);

  Oid        type  = ARR_ELEMTYPE(arr);
  int16      elmlen;
  bool       elmbyval;
  char       elmalign;
  Datum*     elemsp;
  bool*      nullsp;
  int        nelemsp;

  int* idx = (int*)ARR_DATA_PTR(idx_arr);
  if (ARR_NDIM(idx_arr) == 0 ) {
     PG_RETURN_POINTER(construct_empty_array(type));
  }
  const size_t n = ARR_DIMS(idx_arr)[0];
  get_typlenbyvalalign(type, &elmlen, &elmbyval, &elmalign);
  deconstruct_array(arr, type, elmlen, elmbyval, elmalign, &elemsp, &nullsp, &nelemsp);

  Datum* slice = (Datum*)palloc(n * sizeof(Datum));
  for (size_t i = 0; i < n; ++i) {
     int x = idx[i];
     if (x < 0) { x = nelemsp + x + 1; } 
     if (x <= 0 || x > nelemsp) {
       elog(ERROR, "Index entry %d is out of array bounds", idx[i] );
       PG_RETURN_NULL();
     }
     slice[i] = elemsp[x - 1]; //SQL and C covnention is different
  }
  ArrayType* res = construct_array(slice, n, type, elmlen, elmbyval, elmalign);
  PG_RETURN_POINTER(res);
}


extern "C" { PG_FUNCTION_INFO_V1(slice_bit); Datum slice_bit(PG_FUNCTION_ARGS); }

Datum slice_bit(PG_FUNCTION_ARGS) {
  ArrayType* arr = PG_GETARG_ARRAYTYPE_P(0);
  ArrayType* idx_arr = PG_GETARG_ARRAYTYPE_P(1);

  Oid        type  = ARR_ELEMTYPE(arr);
  int16      elmlen;
  bool       elmbyval;
  char       elmalign;
  Datum*     elemsp;
  bool*      nullsp;
  int        nelemsp;

  bool* idx = (bool*)ARR_DATA_PTR(idx_arr);
  if (ARR_NDIM(idx_arr) == 0 ) {
     elog(ERROR, "Can't use empty array as index");
     PG_RETURN_NULL();
  }

  const size_t n = ARR_DIMS(idx_arr)[0];
  get_typlenbyvalalign(type, &elmlen, &elmbyval, &elmalign);
  deconstruct_array(arr, type, elmlen, elmbyval, elmalign, &elemsp, &nullsp, &nelemsp);

  if (n != (size_t)nelemsp) {
    elog(ERROR, "Masking array must be same length as input array");
    PG_RETURN_NULL();
  }
  size_t cnt  = 0;
  for (size_t i = 0; i < n; ++i ) { if (idx[i]) { cnt++; } };

  if (cnt == 0) {
     PG_RETURN_POINTER(construct_empty_array(type));
  }

  Datum* slice = (Datum*)palloc(cnt * sizeof(Datum));
  size_t offset = 0;
  for (size_t i = 0; i < n; ++i) {
     if (idx[i]) {
        slice[offset++] = elemsp[i];
     }
  }
  ArrayType* res = construct_array(slice, cnt, type, elmlen, elmbyval, elmalign);
  PG_RETURN_POINTER(res);
}
