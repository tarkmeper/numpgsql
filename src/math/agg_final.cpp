#include "agg.hpp"

extern "C" {
  PG_FUNCTION_INFO_V1(internal_to_array);
  Datum internal_to_array(PG_FUNCTION_ARGS);
}


Datum internal_to_array(PG_FUNCTION_ARGS) {
  //TODO trigger destructor
  Assert(AggCheckCallContext(fcinfo, NULL));
  if (PG_ARGISNULL(0)) {
     PG_RETURN_NULL();
  } else {
     AggInternal* state = (AggInternal*)(PG_GETARG_POINTER(0));
     PG_RETURN_ARRAYTYPE_P( state->fnc(state)  );
  }
}

