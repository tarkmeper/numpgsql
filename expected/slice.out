-- Integer arrays
SELECT ARRAY[1,2,3,4,5,6] @ ARRAY[1];
 ?column? 
----------
 {1}
(1 row)

SELECT ARRAY[1,2,3,4,5,6] @ ARRAY[-1];
 ?column? 
----------
 {6}
(1 row)

SELECT ARRAY[1,2,3,4,5,6] @ ARRAY[1,1,1,-1,-1];
  ?column?   
-------------
 {1,1,1,6,6}
(1 row)

--Other Types
SELECT ARRAY['asd', 'fds', 'wef'] @ ARRAY[2, -2, 2];
   ?column?    
---------------
 {fds,fds,fds}
(1 row)

-- out of bounds
SELECT ARRAY[1,2,3,4,5,6] @ ARRAY[10];
ERROR:  Index entry 10 is out of array bounds
SELECT ARRAY[1,2,3,4,5,6] @ ARRAY[0];
ERROR:  Index entry 0 is out of array bounds
SELECT ARRAY[1,2,3,4,5,6] @ ARRAY[-7];
ERROR:  Index entry -7 is out of array bounds
